Immediately show the effects of changes to setup code in the play space

balderdash opened the pull request:
The purpose of this PR is to give you immediate feedback when you make changes to the code that affect the initial state of your Game Lab program. That means always showing you any sprites that you create before the draw loop is called (i.e. blocks under `when run` in Game Lab Jr), and possibly anything that happens in the first iteration of your draw loop.

![mittens](https://user-images.githubusercontent.com/1070243/38758763-67cd474c-3f27-11e8-984b-951358d12784.gif)

This is accomplished by running your program as soon as you load the page, and again every time you make a change, but with p5 in a "no loop" state so that it only runs one draw loop. Exactly what code is run in that single iteration of the draw loop is configurable, so it can either be `drawSprites()`, the student's actual `draw()` function, or even a custom snippet of JavaScript.

balderdash commented:
cc @epeach 

This should be a no-op until you enable `auto_run_setup` on a level. It sorta works on gamelab sr, but things get weird when you try typing things in text mode and it attempts to run your half-finished code. Definitely needs more work before being used outside of jr.

balderdash commented:
So apparently it wasn't that hard to call `drawSprites()` alone, but now I'm not sure if I like than more than running one iteration of the draw loop, because it skips the background color, title screen, and any other non-sprite things that we might let you set in `when run`.

joshlory commented:
We may want to make blocks that use randomness be deterministic during predraw 😄 

![random](https://user-images.githubusercontent.com/413693/38397946-16ef6e3c-38f5-11e8-92a5-eb7edec84854.gif)

joshlory commented:
Should the block be "when initialized" or "setup" instead of "when run"?

joshlory commented on apps/src/gamelab/GameLab.js:
```diff
@@ -557,6 +581,43 @@ GameLab.prototype.reset = function (ignore) {
+      const loadingElement = document.createElement('span');
+      loadingElement.id = 'p5_loading';
+      document.body.appendChild(loadingElement);
+    }
```
My initial thought was this looks a bit hacky, but indeed that's how p5.js recommends overriding the loading screen ¯\\\_(ツ)_/¯.

https://github.com/processing/p5.js/blob/edc9c2ef6901e109a328bcc000fbd35a2fed0616/src/core/core.js#L55-L57

joshlory commented on apps/src/gamelab/GameLab.js:
```diff
@@ -557,6 +581,43 @@ GameLab.prototype.reset = function (ignore) {
+    );
+
+    document.body.appendChild(duplicateCanvas);
+    setTimeout(() => document.body.removeChild(duplicateCanvas), 0);
```
I can't decide if this feels like a total hack, or an elegant solution to the redraw problem 😄.  Ideally p5 would take an upstream change to give us more hooks, so we could get flicker-free-redraws without the duplicate canvas.

Another option: can we keep p5 running, clear the canvas, then redraw to avoid the flicker?

islemaster replied:
This is sort of awesome hacky double-buffering :grin: and I'd definitely suggest we use p5 for this ([see `createGraphics`](https://p5js.org/examples/structure-create-graphics.html)) if it's easy to keep an instance around, but I'm not sure it is.  I'm fine with this approach but would suggest some adjustments.

- First, there are some good comments here but for something like this I'd appreciate a larger comment (maybe on the function) describing the high-level approach.
- Even though you've throttled this method, it's going to run a _lot_ so doing some easy optimizations now is probably worth it.
  - Avoid creating elements in this method; let's make one duplicateCanvas and one loadingElement, and reuse them.
  - Even cutting DOM queries - we can probably safely cache a reference to the visualizationCanvas.

joshlory commented on apps/src/gamelab/GameLabJr.interpreted.js:
```diff
@@ -256,68 +256,71 @@ function hideTitleScreen() {
-    const param = inputEvents[i].param;
-    if (eventType(param)) {
-      event();
+  if (World.frameCount > 1) {
```
Is this still needed if we're only calling `drawSprites()` for predraw?

balderdash replied:
I'm in favor of running the draw loop instead, but I made it an levelbuilder setting so we can try both. I'm going to leave this in unless/until we settle on using `drawSprites()` (or perhaps defining a `drawWorld()` to handle the actual rendering part of the draw loop and calling that instead).

joshlory commented on apps/src/gamelab/GameLabP5.js:
```diff
@@ -38,6 +38,19 @@ var GameLabP5 = function () {
+      this.p5._loop = true;
+    } else {
+      this.p5.noLoop();
+    }
```
Not sure I'm following what's going on here — why is `setLoop` needed?

balderdash replied:
This is just to encapsulate the logic of putting p5 into loop/noLoop mode when we switch between running setup code and running the full program, since it's not quite as straightforward as calling `loop()` and `noLoop()`.

islemaster commented on apps/src/gamelab/GameLab.js:
```diff
@@ -281,6 +281,19 @@ GameLab.prototype.init = function (config) {
+          this.runSetupCode.bind(this));
+      } else {
+        this.studioApp_.editor.on('change', this.runSetupCode.bind(this));
+        // Droplet doesn't automatically bubble up aceEditor changes
```
:sob:

islemaster commented on apps/src/gamelab/GameLab.js:
```diff
@@ -281,6 +281,19 @@ GameLab.prototype.init = function (config) {
+        this.studioApp_.editor.on('change', this.runSetupCode.bind(this));
+        // Droplet doesn't automatically bubble up aceEditor changes
+        this.studioApp_.editor.aceEditor.on('change',
+          this.runSetupCode.bind(this));
```
Minor nit: Creating two bound functions when only one is needed. :trollface:

islemaster commented on apps/src/gamelab/GameLab.js:
```diff
@@ -503,11 +516,22 @@ GameLab.prototype.startTickTimer = function () {
  */
-GameLab.prototype.reset = function (ignore) {
+GameLab.prototype.resetHandler = function (ignore) {
+  if (this.level.autoRunSetup && !this.level.edit_blocks) {
```
You check this compound condition in at least two places.  Maybe worth extracting a helper.

islemaster commented on apps/src/gamelab/GameLab.js:
```diff
@@ -557,6 +581,43 @@ GameLab.prototype.reset = function (ignore) {
+    }
+  }
+  this.execute(false /* keep Ticking */);
+}, 50, {leading: true, maxWait: 250});
```
Usually when I see [`_.debounce`](https://lodash.com/docs/4.17.5#debounce) with `leading: true` and a `maxWait` I suspect what you actually want is [`_.throttle`](https://lodash.com/docs/4.17.5#throttle).  The behavior will be slightly different, but it seems like what you're trying to achieve is "Don't call this more than once every 50ms" which is what `_.throttle` is designed for.

balderdash commented:
Alright, I made a few changes.

* I can keep p5 running, so each run is faster and there's no more flickering
  * Unfortunately this means my elegant double-buffering code is no longer necessary
* Added options to levelbuilder to customize what runs in the single iteration of the draw loop (`draw()` vs `drawSprites()` vs something else entirely)
* Removed the `debounce` now that it's so much faster. There are still some issues with syntax error highlighting in droplet text mode, but it mostly works
* I tried to make the naming a bit more consistent, but variable names are hard

@joshlory and @islemaster PTAL

joshlory commented on apps/src/gamelab/GameLabJr.interpreted.js:
```diff
@@ -256,68 +256,71 @@ function hideTitleScreen() {
-        sprite.things_to_say.shift();
-      } else if (sprite.things_to_say[0][1] > -1) {
-        sprite.things_to_say[0][1]--;
+      // Make sprites say things
```
[Out of scope for this PR] Still used in the curriculum?

balderdash replied:
As far as I've seen, it's not in any of our planned lessons. I imagine we'll still want to support it for freeplay, with at least the same level of personal info filtering we have for playlab.

joshlory commented on lib/cdo/shared_constants.rb:
```diff
@@ -48,6 +48,15 @@ module SharedConstants
+  GAMELAB_AUTORUN_OPTIONS = OpenStruct.new(
+    {
+      draw_loop: 'DRAW_LOOP',
+      draw_sprites: 'DRAWP_SPRITES',
```
Nit: typo.

joshlory commented on dashboard/app/views/levels/editors/_gamelab.html.haml:
```diff
@@ -80,6 +80,35 @@
+    ], selected: @level.auto_run_setup)
+  = selector
+  = f.label 'Custom setup code'
+  = f.text_field :custom_setup_code
```
This makes me a bit nervous.  While the Game Lab Sr. API set is pretty well fixed, we'll still be making changes to Jr. API.  There's a potential here to have lots of untested, unreviewed code written in levels that we depend on in shipped courses.

I don't have a better solution...

balderdash replied:
I'd like to have some way to warn us about levels in "real" scripts that have custom helper code, custom blocks, or anything else like that. Would it be too heavy-handed to make such levels fail validation?

islemaster commented on dashboard/app/views/levels/editors/_gamelab.html.haml:
```diff
@@ -80,6 +80,35 @@
+      }
+      autoRunSetup.onchange = changeHandler;
+      changeHandler();
+    })();
```
JavaScript in HAML?  When [levelbuilder_gamelab.js](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/sites/studio/pages/levelbuilder_gamelab.js) already exists?

![boo](https://user-images.githubusercontent.com/1615761/38954270-11a67d56-4306-11e8-9282-e37c0f17d00c.gif)

balderdash replied:
I am ashamed

balderdash commented:
![](https://github.trello.services/images/mini-trello-icon.png) [GLJ: Render code under "when run" even before run is clicked](https://trello.com/c/IB9N6rVn/85-glj-render-code-under-when-run-even-before-run-is-clicked)
